File: Python/Baekjoon/15000/15060.py
from sys import stdin
from collections import deque

N, R = map(int, stdin.readline().rstrip().split())
pr = [i for i in range(N + 1)]
rt, df = 0, 0


def find(i):
    if pr[i] != i:
        pr[i] = find(pr[i])
    return pr[i]


def uni(i, j):
    i, j = find(i), find(j)
    pr[max(i, j)] = min(i, j)


money = [[0 for j in range(N + 1)] for i in range(N + 1)]
arr = []
pw = []

for i in range(18):
    pw.append(1 << i)

for i in range(R):
    A, B, C = map(int, stdin.readline().rstrip().split())
    money[A][B] = C
    arr.append([C, A, B])

arr = list(sorted(arr))
child = [[] for i in range(N + 1)]
for C, A, B in arr:
    if find(A) == find(B):
        continue

    uni(find(A), find(B))
    df += C
    child[A].append([B, C])
    child[B].append([A, C])

    if not rt:
        rt = A

st = deque([rt])
checked = [False for i in range(N + 1)]
checked[rt] = True

depth = [0 for i in range(N + 1)]
parent = [[[0, 0] for j in range(18)] for i in range(N + 1)]

while len(st):
    p = st.pop()
    for A, B in child[p]:
        if checked[A]:
            continue

        checked[A] = True
        parent[A][0] = [p, B]
        depth[A] = depth[p] + 1

        K = 1
        while 1 << K <= N:
            if not parent[A][K - 1][0]:
                break
            parent[A][K] = [
                parent[parent[A][K - 1][0]][K - 1][0],
                max(parent[parent[A][K - 1][0]][K - 1][1], parent[A][K - 1][1]),
            ]
            K += 1

        st.append(A)


Q = int(stdin.readline())
for i in range(Q):
    A, B = map(int, stdin.readline().rstrip().split())
    base, res = money[A][B] + df, 0

    if depth[B] > depth[A]:
        A, B = B, A

    diff = depth[A] - depth[B]
    for j in range(17, -1, -1):
        if pw[j] <= diff:
            diff -= pw[j]
            res = max(res, parent[A][j][1])
            A = parent[A][j][0]

    for j in range(17, -1, -1):
        if parent[A][j][0] != parent[B][j][0]:
            res = max(res, parent[A][j][1], parent[B][j][1])
            A, B = parent[A][j][0], parent[B][j][0]

    if A != B:
        res = max(res, parent[A][0][1], parent[B][0][1])
    print(base - res)

File: C/Baekjoon/15000/15060.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <unordered_map>
#include <stack>
#define MAX 100001
using namespace std;

int parent[MAX];

int find(int i)
{
    if (i != parent[i])
        parent[i] = find(parent[i]);
    return parent[i];
}

void uni(int i, int j)
{
    i = find(i), j = find(j);
    parent[max(i, j)] = min(i, j);
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    int N, M, Q, U, V, A, B, rt = -1, W, def = 0;
    vector<pair<int, pair<int, int>>> lines;

    cin >> N >> M;
    for (int i = 1; i <= N; i++)
        parent[i] = i;

    unordered_map<int, int> money[N + 1];
    while (M--)
    {
        cin >> U >> V >> W;
        money[U][V] = W;
        money[V][U] = W;

        lines.push_back({W, {U, V}});
    }
    sort(lines.begin(), lines.end());

    vector<pair<int, int>> arr[N + 1];
    int depth[N + 1] = {};
    stack<int> st;
    pair<int, int> parent[N + 1][18] = {(pair<int, int>){0, 0}};
    bool checked[N + 1] = {};

    for (pair<int, pair<int, int>> i : lines)
    {
        A = find(i.second.first), B = find(i.second.second);
        if (A == B)
            continue;

        uni(A, B);
        def += i.first;

        arr[i.second.first].push_back({i.second.second, i.first});
        arr[i.second.second].push_back({i.second.first, i.first});

        if (rt == -1)
            rt = A;
    }

    st.push(rt);
    checked[rt] = true;

    while (!st.empty())
    {
        A = st.top();
        st.pop();

        for (pair<int, int> i : arr[A])
        {
            if (checked[i.first])
                continue;

            parent[i.first][0] = {A, i.second};
            checked[i.first] = true;
            depth[i.first] = depth[A] + 1;

            for (int j = 1; 1 << j <= N; j++)
            {
                if (!parent[i.first][j - 1].first)
                    break;

                parent[i.first][j] = {
                    parent[parent[i.first][j - 1].first][j - 1].first,
                    max(parent[parent[i.first][j - 1].first][j - 1].second, parent[i.first][j - 1].second)};
            }

            st.push(i.first);
        }
    }

    cin >> Q;
    while (Q--)
    {
        cin >> A >> B;
        if (depth[A] < depth[B])
            swap(A, B);

        int diff = depth[A] - depth[B], i = 0, U = money[A][B];
        int res = 0;

        while (diff)
        {
            if (diff % 2)
            {
                res = max(res, parent[A][i].second);
                A = parent[A][i].first;
            }
            i++;
            diff >>= 1;
        }

        for (int i = 17; i >= 0; i--)
        {
            if (parent[A][i].first != parent[B][i].first)
            {
                res = max(res, max(parent[A][i].second, parent[B][i].second));
                A = parent[A][i].first, B = parent[B][i].first;
            }
        }

        if (A != B)
            res = max(res, max(parent[A][0].second, parent[B][0].second));
        cout << def - res + U << '\n';
    }
}